src/klein/app.py

```diff
@@ -198,16 +200,33 @@ def index(request):
 
         @returns: decorated handler function.
         """
-        segment_count = self._segments_in_url(url) + self._subroute_segments
 
-        @named("router for '" + url + "'")
+        def _urlToString(_url):
+            return _url.asText()
+
+        if isinstance(url, string_types + (binary_type, )):
+            url = URL.fromText(url)
```

Comment:
`URL` will (perhaps unfortunately) take bytes as input without complaining, but it will be wrong.

```
>>> URL.fromText(b"/foo").asText()
"b'/foo'"
```

Looks like it'll take any type:

```
>>> URL.fromText(object()).asText()
'%3Cobject%20object%20at%200x100538160%3E'
```

But obviously, it's calling `str` on it, which is not awesome in the above examples.

If you want to allow bytes as input here, you should add a test for that.